data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java

```diff
@@ -731,13 +731,9 @@ List<Withdrawal> getExpectedWithdrawalsFromState(
   }
 
   public SafeFuture<Optional<Bytes32>> getFinalizedBlockRoot(final UInt64 slot) {
-    final SafeFuture<Optional<SignedBeaconBlock>> futureFinalizedBlock =
-        combinedChainDataClient.getFinalizedBlockInEffectAtSlot(slot);
-    return futureFinalizedBlock.thenApply(
-        maybeFinalizedBlock ->
-            maybeFinalizedBlock
-                .filter(block -> block.getSlot().equals(slot))
-                .map(SignedBeaconBlock::getRoot));
+    return combinedChainDataClient
+        .getFinalizedBlockAtSlotExact(slot)
+        .thenApply(maybeFinalizedBlock -> maybeFinalizedBlock.map(SignedBeaconBlock::getRoot));
```

Comment:
we'd need to be careful if we go down changing these low level functions, hopefully theres god tests...

 - `*atSlotExact` is 'if theres a block at this slot exactly, then i want a block, otherwise i want empty
 - `*inEffectAtSlot` means, i want whatever is the 'current' block at a given slot, so if im at slot N, and there was no block at slot N, but there was one at slot N-1, then my expectation is that i get the block at slot N-1.

By this reasoning, i'm fairly sure that we don't want to make this specific change proposed...

Reply:
we were filtering it lines below:
```
                .filter(block -> block.getSlot().equals(slot))
```
we need exact slot in this function. I understand the difference

Reply:
actually I've broken tests, I don't understand something here, rechecking

Reply:
RecentChainData.getBlockRootBySlot was actually in effect, renamed for clarity, fixed tests